### PR TITLE
Serialized music controls

### PR DESCRIPTION
**Pull request by dgreenhalgh**
Serialized/deserialized messages for Play, Pause, and Skip controls. Also clarified some unit testing procedure and wrote unit tests for music control serialization

### CODE REVIEW

**Comment by theJenix** [comment]
Talked to David, and there are 2 changes coming.  One: split the unit tests into files that match the code under test (e.g. PlayMessageTest for tests for PlayMessage), and two: migrate boilerplate test code to a base class, to make building message unit tests easier to write.  Once these are done, I think this is good to pull in.

**Comment by reidbaker** [comment]
Did we decide to do android unit tests or java unit tests?

File: docs/unitTesting.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#Unit testing
+
+Because Eclipse can be weird, here is how **non-Android** unit tests in fanClub-test must be run:
+* Write non-Android style regular unit tests
+* In your run configurations, make sure to use the Android JUnit test launcher
+* Run your unit tests as normal JUnit tests
+
+*I wrote this so that other people don't spend two days debugging only to find that you missed a single configurable.*
+***THIS IS MY STORY***
```
Comment (reidbaker) [inline docs/unitTesting.md]: Lets drop THIS IS MY STORY

File: fanClub-test/src/com/lastcrusade/fanclub/message/MessengerTest.java [hunk 2]
```diff
@@ -12,11 +12,18 @@
...
 import org.junit.Test;
 
+import com.lastcrusade.fanclub.LandingActivity;
+
+/**
+ * Doesn't inherit from SerializationTest and is separate from its generalized
+ * test structure. Meaning, the code repetition between this class and
+ * SerializationTest is necessary!
+ */
 public class MessengerTest {
```
Comment (reidbaker) [inline fanClub-test/src/com/lastcrusade/fanclub/message/MessengerTest.java:23]: nitpick: Can @Test line up with public

File: fanClub-test/src/com/lastcrusade/fanclub/message/PlayMessageTest.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import static org.junit.Assert.assertTrue;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import org.junit.Test;
+
+public class PlayMessageTest extends SerializationTest<PlayMessage> {
+	
+	@Test
```
Comment (reidbaker) [inline fanClub-test/src/com/lastcrusade/fanclub/message/PlayMessageTest.java]: nitpick: Move @test to line up with public and indent both `PlayMessage` and `assertEquals`

File: fanClub-test/src/com/lastcrusade/fanclub/message/SerializationTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+
+import org.junit.Test;
+
+public class SerializationTest<T extends IMessage> {
+	
+	public T testSerializeMessage(T message) throws IOException {
```
Comment (reidbaker) [inline fanClub-test/src/com/lastcrusade/fanclub/message/SerializationTest.java]: nitpick: can lines 18-34 be indented at the same depth.
If they look ok in your code check for tabs

File: fanClub-test/src/com/lastcrusade/fanclub/message/SerializationTest.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        IMessage rcvMessage = rcvMessenger.getReceivedMessage();
+        assertNotNull(rcvMessage);
+        assertEquals(message.getClass(), rcvMessage.getClass());
+        
+        //make sure all bytes are consumed
+        assertEquals(0, is.available());
+        return (T)rcvMessage;
+	}
+	
+	public T testDeserializeMessage(T message) throws IOException {
```
Comment (reidbaker) [inline fanClub-test/src/com/lastcrusade/fanclub/message/SerializationTest.java]: nitpick: can lines 38-53 be indented at the same depth.
If they look ok in your code check for tabs

File: fanClub-test/src/com/lastcrusade/fanclub/message/SerializationTest.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        IMessage rcvMessage = messenger.getReceivedMessage();
+        assertNotNull(rcvMessage);
+        assertEquals(message.getClass(), rcvMessage.getClass());
+        
+        //make sure all bytes are consumed
+        assertEquals(0, bais.available());
+        return (T)rcvMessage;
+	}
+	
+	private InputStream simulateSendAndReceive(
```
Comment (reidbaker) [inline fanClub-test/src/com/lastcrusade/fanclub/message/SerializationTest.java]: nitpick: can lines 57 and 58 be indented at the same depth.
If they look ok in your code check for tabs

File: fanClub/src/com/lastcrusade/fanclub/message/PauseMessage.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+public class PauseMessage implements IMessage {
+	private String string;
+	private final String PAUSE_MESSAGE = "Pause";
+
+	@Override
+	public void deserialize(InputStream input) throws IOException {
+		byte[] bytes = new byte[1024];
+		int read = 0;
+		ByteArrayOutputStream out = new ByteArrayOutputStream();
+		while((read = input.read(bytes)) > 0)
```
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/message/PauseMessage.java]: Can you put brackets {} on the `while`?

File: fanClub/src/com/lastcrusade/fanclub/message/PlayMessage.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+public class PlayMessage implements IMessage {
+	private String string;
+	private final String PLAY_MESSAGE = "Play"; 
+
+	@Override
+	public void deserialize(InputStream input) throws IOException {
+		byte[] bytes = new byte[1024];
+		int read = 0;
+		ByteArrayOutputStream out = new ByteArrayOutputStream();
+		while((read = input.read(bytes)) > 0)
```
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/message/PlayMessage.java]: Can you put brackets {} on the `while`?

File: fanClub/src/com/lastcrusade/fanclub/message/SkipMessage.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+public class SkipMessage implements IMessage {
+	private String string;
+	private final String SKIP_MESSAGE= "Skip";
+
+	@Override
+	public void deserialize(InputStream input) throws IOException {
+		byte[] bytes = new byte[1024];
+		int read = 0;
+		ByteArrayOutputStream out = new ByteArrayOutputStream();
+		while((read = input.read(bytes)) > 0)
```
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/message/SkipMessage.java]: Can you put brackets {} on the `while`?

**Comment by dgreenhalgh** [comment]
Almost outstanding issues on this pull request have to do with tabs/spaces.

In Eclipse, you can control tabs/spaces in your general editor and in your Android editor, but not your Java editor. Unfortunately, the Java editor is inserting spaces for tabs and overriding the general editor.  That being said, while this sucks, it should only be an issue in our JUnit tests because they're technically written in Java rather than Android.

**Comment by dgreenhalgh** [comment]
I was able to replace tabs with spaces using sed, Every issue in this pull request has been addressed and fixed, so we should be ready to pull this in

**Comment by reidbaker** [comment]
looks great I am pulling it in.

**Pull request merged**